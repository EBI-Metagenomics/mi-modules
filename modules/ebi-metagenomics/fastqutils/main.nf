
process FASTQUTILS {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/fastq_utils:0.25.2--h50ea8bc_1':
        'biocontainers/fastq_utils:0.25.2--h50ea8bc_1' }"

    input:
    tuple val(meta), path(fastq)

    output:
    tuple val(meta), path("*.txt"), emit: sanity
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    fastq_info \\
        ${fastq} \\
        2>&1 >/dev/null \\
        | tail -1 \\
        > ${prefix}_sanity.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqutils: \$(fastq_info 2>&1 >/dev/null | head -1 | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_sanity.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqutils: \$(fastq_info 2>&1 >/dev/null | head -1 | cut -d' ' -f2)
    END_VERSIONS
    """
}
