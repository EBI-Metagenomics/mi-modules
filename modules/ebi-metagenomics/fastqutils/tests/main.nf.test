
nextflow_process {

    name "Test Process FASTQUTILS"
    script "../main.nf"
    process "FASTQUTILS"

    tag "modules"
    tag "modules_nfcore"
    tag "fastqutils"

    test("fastqutils - noerrors_pe") {
        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_trim_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_trim_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("fastqutils - brokenfileerror") {
        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_test_brokenfile_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_test_brokenfile_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("fastqutils - headermatcherror") {
        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_test_headermatch_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_test_headermatch_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("fastqutils - headeruniqueerror") {
        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_test_headerunique_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_test_headerunique_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("fastqutils - readcounterror") {
        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_test_readcount_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_test_readcount_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("fastqutils - stub") {

        options "-stub"

        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ file("${moduleDir}/tests/fixtures/ERR4674038_trim_1.fastq", checkIfExists: true),
                     file("${moduleDir}/tests/fixtures/ERR4674038_trim_2.fastq", checkIfExists: true)
                    ]
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
