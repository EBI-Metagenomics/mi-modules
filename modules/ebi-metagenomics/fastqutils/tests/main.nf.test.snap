{
    "fastqutils - noerrors_pe": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,d36f8f9425c4a8000ad9c4a97185aca5"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,d36f8f9425c4a8000ad9c4a97185aca5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:19:09.857435291"
    },
    "fastqutils - brokenfileerror": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,98c83b763d170c0f16bf9000869bd86a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,98c83b763d170c0f16bf9000869bd86a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:39:34.912369382"
    },
    "fastqutils - headeruniqueerror": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,133ae8181370b213d9da44eef0e2421e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,133ae8181370b213d9da44eef0e2421e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:42:30.465915574"
    },
    "fastqutils - readcounterror": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,0437b4c2d398aee0614f0e46235fdc41"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,0437b4c2d398aee0614f0e46235fdc41"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:42:39.557786664"
    },
    "fastqutils - headermatcherror": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,c39c6444668ee1ef02cbc995708018f9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,c39c6444668ee1ef02cbc995708018f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:42:15.58774212"
    },
    "fastqutils - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ],
                "sanity": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_sanity.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a5c6b30aead5ce782fac3c3f0a292704"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-08-01T16:21:48.395516095"
    }
}